New command: rename SharePoint Framework project

aakashbhardwaj619 opened the pull request:
New command: `spfx project rename` to rename SharePoint Framework project. Implements #1349

aakashbhardwaj619 commented:
> Rather than adding your own test project, could you instead reference the one from spfx project upgrade? That will help us to avoid duplication. We'll update the project structure and move the test projects a level up to make it clear that they can be used by any command. Sorry for the trouble.

No problem. I've removed the duplicate test project now and referenced the one in project upgrade.

waldekmastykarz commented:
Awesome! Sorry for the trouble and thank you for such a quick turnaround. We'll review the PR as soon as possible. Looking forward to it! 👍

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -0,0 +1,193 @@
+
+  public getTelemetryProperties(args: CommandArgs): any {
+    const telemetryProps: any = super.getTelemetryProperties(args);
+    telemetryProps.newName = (!(!args.options.newName)).toString();
```
Since `newName` is required, there is no point in tracking its usage because it would always show up as specified

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -0,0 +1,193 @@
+}
+
+class SpfxProjectRenameCommand extends BaseProjectCommand {
+  public constructor() {
```
No need to implement the constructor given we're not doing anything special here and the parent constructor will be called automatically.

aakashbhardwaj619 replied:
Agreed. I've removed the constructor now.

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -0,0 +1,193 @@
+      cmd.log('DONE');
+    }
+
+    cb(`SharePoint Framework project successfully renamed to '${args.options.newName}'`);
```
When the command succeeds, we need to call `cb()` without any arguments. When you pass an argument, it's treated as an error message.

aakashbhardwaj619 replied:
Thanks. Removed the argument.

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    filePaths.forEach((filePath) => {
+      replaceFileContent(filePath);
```
Since each file that needs to be updates has a different structure, why call the same method on all of them? It would be more efficient and reliable to apply only the modifications relevant to the particular file and avoid any possible side-effects.

aakashbhardwaj619 replied:
Hi @waldekmastykarz I updated the **replaceFileContent** method structure to something like this. Let me know if this works:
```
const replaceFileContent = (filePath: string) => {
  if (!fs.existsSync(filePath)) {
    return;
  }
  const existingContent = fs.readFileSync(filePath, 'utf-8');
  let updatedContent;
  if (filePath.endsWith('.json')) {
    updatedContent = JSON.parse(existingContent);

    if (filePath.endsWith('package.json')) {
      //Update specific json properties
    }
    if (filePath.endsWith('.yo-rc.json')) {
      //Update specific json properties
    }
    if (filePath.endsWith('config/package-solution.json')) {
      //Update specific json properties
    }
    if (filePath.endsWith('config/deploy-azure-storage.json')) {
      //Update specific json properties
    }
    fs.writeFileSync(filePath, JSON.stringify(updatedContent, null, 2), 'utf-8');
  }
  if (filePath.endsWith('README.md')) {
    //Update content
  }
  if (this.debug) {
    cmd.log(`Updated ${filePath.split('/').pop()}`);
  }
}
```

aakashbhardwaj619 replied:
Hi @waldekmastykarz I've made all the requested changes along with the above mentioned one. Let me know if this works or any other changes are required. Thanks for your guidance! 😊

waldekmastykarz replied:
Thank you! We'll take a look as quickly as possible 👍

waldekmastykarz replied:
Hey @aakashbhardwaj619, what is the benefit of having one `replaceFileContent` method that uses separate conditional blocks for the different files it's called on, instead of separate methods for each file? It seems like the only common logic is reading file contents, parsing to JSON and writing the update structure to the file, so that could be separated out, but I think that keeping the different modifications separate would make it easier to maintain the code over time.

aakashbhardwaj619 replied:
Hi @waldekmastykarz you are right, no other benefit apart from the common logic you mentioned. I have now created separate methods for updating each of the files. So sorry that you had to review this PR multiple times.

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -0,0 +1,193 @@
+    ];
+
+    const replaceFileContent = (filePath: string) => {
+      if (fs.existsSync(filePath)) {
```
Rather than wrapping the whole block in this `if` statement, you could invert it and simplify the whole block:

```ts
if (!fs.existsSync(filePath)) {
  return;
}

// ... rest of the code
```

aakashbhardwaj619 replied:
Thanks. Updated now.

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -0,0 +1,193 @@
+
+    const replaceFileContent = (filePath: string) => {
+      if (fs.existsSync(filePath)) {
+        let existingContent = fs.readFileSync(filePath, 'utf-8');
```
Since existingContent is not reassigned, it should be defined as `const`

aakashbhardwaj619 replied:
Updated to const now.

waldekmastykarz commented on docs/manual/docs/cmd/spfx/project/project-rename.md:
```diff
@@ -0,0 +1,39 @@
+# spfx project rename
+
+Renames SharePoint Framework project
```
Doesn't match description from the command `Rename SharePoint Framework project` (rename instead of renames)

aakashbhardwaj619 replied:
In the issue #1349 description is mentioned as `Renames SharePoint Framework project`. Please confirm if this should be updated to `Rename SharePoint Framework project`

aakashbhardwaj619 replied:
Got it now. Updated all references of `Rename SharePoint Framework project` to `Renames SharePoint Framework project` to match the issue description.

waldekmastykarz commented on docs/manual/docs/cmd/spfx/project/project-rename.md:
```diff
@@ -0,0 +1,39 @@
+`--verbose`|Runs command with verbose logging
+`--debug`|Runs command with debug logging
+
+!!! important
```
Important block missing in the command help

aakashbhardwaj619 replied:
Added the important block.

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.spec.ts:
```diff
@@ -0,0 +1,413 @@
+
+  it('replaces project name in package.json', (done) => {
+    sinon.stub(command as any, 'getProjectRoot').callsFake(_ => path.join(process.cwd(), projectPath));
+    let replacedContent = `{
```
This should be const

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.spec.ts:
```diff
@@ -0,0 +1,413 @@
+  it('replaces only project name in .yo-rc.json when --generateNewId is not passed', (done) => {
+    sinon.stub(command as any, 'getProjectRoot').callsFake(_ => path.join(process.cwd(), projectPath));
+
+    let replacedContent = `{
```
This should be const

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.spec.ts:
```diff
@@ -0,0 +1,413 @@
+  }
+}`;
+
+    const writeFileSyncSpy: sinon.SinonStub = sinon.stub(fs, 'writeFileSync').callsFake(() => { });
```
To avoid repeating this line in each test, you can move it to the `beforeEach` block on top

aakashbhardwaj619 replied:
Thanks. Updated now.

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.spec.ts:
```diff
@@ -0,0 +1,413 @@
+
+    cmdInstance.action({ options: { newName: 'spfx-react' } }, (err?: any) => {
+      try {
+        assert.equal(writeFileSyncSpy.getCall(0).args[1], replacedContent);
```
Rather than expecting updating the file to be an n-th call, you could call `writeFileSyncSpy.calledWith` and pass the expected arguments. That way, should the implementation change and something else was written before, the order won't matter and it won't break the test

aakashbhardwaj619 replied:
Thanks. Getting used to Sinon! 😃

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.spec.ts:
```diff
@@ -0,0 +1,413 @@
+      return '69cb6882-acc1-4148-b059-31ae149ba077'
+    });
+
+    let replacedContent = `{
```
This should be const

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.spec.ts:
```diff
@@ -0,0 +1,413 @@
+  it('replaces only project name in package-solution.json when --generateNewId is not passed', (done) => {
+    sinon.stub(command as any, 'getProjectRoot').callsFake(_ => path.join(process.cwd(), projectPath));
+
+    let replacedContent = `{
```
This should be const

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.spec.ts:
```diff
@@ -0,0 +1,413 @@
+      return '69cb6882-acc1-4148-b059-31ae149ba077'
+    });
+
+    let replacedContent = `{
```
This should be const

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.spec.ts:
```diff
@@ -0,0 +1,413 @@
+  it('replaces project name in deploy-azure-storage.json', (done) => {
+    sinon.stub(command as any, 'getProjectRoot').callsFake(_ => path.join(process.cwd(), projectPath));
+
+    let replacedContent = `{
```
This should be const

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.spec.ts:
```diff
@@ -0,0 +1,413 @@
+  it('replaces project name in README.md', (done) => {
+    sinon.stub(command as any, 'getProjectRoot').callsFake(_ => path.join(process.cwd(), projectPath));
+
+    let replacedContent = `## spfx-react
```
This should be const

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.spec.ts:
```diff
@@ -106,9 +106,21 @@ describe(commands.PROJECT_RENAME, () => {
+        path: projectPath,
+        packageJson: {
+          dependencies: {},
+          name: projectPath.split('/').pop()
```
Rather than using fixed `/` we should use OS-specific path separator (`path.sep`). In this particular case, to get the file-/directory name from the path, it would be better to use `path.basename` (https://nodejs.org/api/path.html#path_path_basename_path_ext)

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -90,81 +90,101 @@ class SpfxProjectRenameCommand extends BaseProjectCommand {
+      }
+      fs.writeFileSync(filePath, JSON.stringify(updatedContent, null, 2), 'utf-8');
+      if (this.debug) {
+        cmd.log(`Updated ${filePath.split('/').pop()}`);
```
Rather than using fixed `/` we should use OS-specific path separator (`path.sep`). In this particular case, to get the file-/directory name from the path, it would be better to use `path.basename` (https://nodejs.org/api/path.html#path_path_basename_path_ext)

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -90,81 +90,101 @@ class SpfxProjectRenameCommand extends BaseProjectCommand {
+      }
+      fs.writeFileSync(filePath, JSON.stringify(updatedContent, null, 2), 'utf-8');
+      if (this.debug) {
+        cmd.log(`Updated ${filePath.split('/').pop()}`);
```
Rather than using fixed `/` we should use OS-specific path separator (`path.sep`). In this particular case, to get the file-/directory name from the path, it would be better to use `path.basename` (https://nodejs.org/api/path.html#path_path_basename_path_ext)

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -90,81 +90,101 @@ class SpfxProjectRenameCommand extends BaseProjectCommand {
+      }
+      fs.writeFileSync(filePath, JSON.stringify(updatedContent, null, 2), 'utf-8');
+      if (this.debug) {
+        cmd.log(`Updated ${filePath.split('/').pop()}`);
```
Rather than using fixed `/` we should use OS-specific path separator (`path.sep`). In this particular case, to get the file-/directory name from the path, it would be better to use `path.basename` (https://nodejs.org/api/path.html#path_path_basename_path_ext)

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -90,81 +90,101 @@ class SpfxProjectRenameCommand extends BaseProjectCommand {
+      }
+      fs.writeFileSync(filePath, JSON.stringify(updatedContent, null, 2), 'utf-8');
+      if (this.debug) {
+        cmd.log(`Updated ${filePath.split('/').pop()}`);
```
Rather than using fixed `/` we should use OS-specific path separator (`path.sep`). In this particular case, to get the file-/directory name from the path, it would be better to use `path.basename` (https://nodejs.org/api/path.html#path_path_basename_path_ext)

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -90,81 +90,101 @@ class SpfxProjectRenameCommand extends BaseProjectCommand {
+      const updatedContent = existingContent.replace(new RegExp(projectName, 'g'), args.options.newName);
+      fs.writeFileSync(filePath, updatedContent, 'utf-8');
       if (this.debug) {
         cmd.log(`Updated ${filePath.split('/').pop()}`);
```
Rather than using fixed `/` we should use OS-specific path separator (`path.sep`). In this particular case, to get the file-/directory name from the path, it would be better to use `path.basename` (https://nodejs.org/api/path.html#path_path_basename_path_ext)

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -174,9 +194,11 @@ class SpfxProjectRenameCommand extends BaseProjectCommand {
-    });
+    replacePackageJsonContent(path.join(this.projectRootPath, 'package.json'));
+    replaceYoRcJsonContent(path.join(this.projectRootPath, '.yo-rc.json'));
+    replacePackageSolutionJsonContent(path.join(this.projectRootPath, 'config/package-solution.json'));
```
Rather than using a fixed `/`, we should specify all segments separately and have Node join the path properly following the separator for the current OS, eg `path.join(this.projectRootPath, 'config', 'package-solution.json')`

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -174,9 +194,11 @@ class SpfxProjectRenameCommand extends BaseProjectCommand {
+    replacePackageJsonContent(path.join(this.projectRootPath, 'package.json'));
+    replaceYoRcJsonContent(path.join(this.projectRootPath, '.yo-rc.json'));
+    replacePackageSolutionJsonContent(path.join(this.projectRootPath, 'config/package-solution.json'));
+    replaceDeployAzureStorageJsonContent(path.join(this.projectRootPath, 'config/deploy-azure-storage.json'));
```
Rather than using a fixed `/`, we should specify all segments separately and have Node join the path properly following the separator for the current OS, eg `path.join(this.projectRootPath, 'config', 'deploy-azure-storage.json')`

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -90,81 +90,101 @@ class SpfxProjectRenameCommand extends BaseProjectCommand {
-      path.join(this.projectRootPath, 'config/deploy-azure-storage.json'),
-      path.join(this.projectRootPath, 'README.md')
-    ];
+    const replacePackageJsonContent = (filePath: string) => {
```
Let's move these functions outside of `action` to a separate class-level method.

aakashbhardwaj619 replied:
Hi @waldekmastykarz! Shall I create a single class method that contains all these functions or do each of these functions need to be written as separate class methods?

waldekmastykarz replied:
Let's do a separate class method for each modification. As the project evolves and new elements are added it will be easier to keep track of the different pieces that need to be renamed.

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -90,81 +90,101 @@ class SpfxProjectRenameCommand extends BaseProjectCommand {
+      if (!fs.existsSync(filePath)) {
+        return;
+      }
+      const existingContent = fs.readFileSync(filePath, 'utf-8');
```
To avoid the CLI crash on exceptions, let's wrap reading file contents, deserializing JSON and writing the file in a `try..catch` block and break the command execution on an exception

aakashbhardwaj619 replied:
@waldekmastykarz Just to be clear, the command execution should break with an error message if there is an exception in updating any of these files?

waldekmastykarz replied:
Yes, so rather than returning an unhandled exception, we should call `cb()` (the action callback) passing the error message from the exception wrapped in `CommandError` as its argument.

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -90,81 +90,101 @@ class SpfxProjectRenameCommand extends BaseProjectCommand {
+        updatedContent.name) {
+        updatedContent.name = args.options.newName;
+      }
+      fs.writeFileSync(filePath, JSON.stringify(updatedContent, null, 2), 'utf-8');
```
If the file contents have not been changed, we shouldn't write the file back to the disk

garrytrinder commented:
This pull request has been converted to `draft` to signify that it is still a work in progress as there are outstanding changes requested following a pull request review. 

After making the required changes and when you're ready to get feedback on your pull request, you can mark your draft pull request as ready for review.

More Information 
 - [Changing the stage of a pull request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request)

waldekmastykarz commented:
Hey @aakashbhardwaj619, did you have a chance to go through the latest findings so that we can proceed with processing the PR and getting it merged?

aakashbhardwaj619 commented:
Hi @waldekmastykarz I am done with updating the code as per review comments but facing some issues with tests after the updated structure. I couldn't spend much time on it lately due to some other priorities but will try to fix and commit in 1-2 days.

waldekmastykarz commented:
Awesome! Appreciate your help, and don't hesitate to reach out if we can help with anything

aakashbhardwaj619 commented:
Hi @waldekmastykarz, I'm having trouble in breaking the command execution when error occurs in the replace method for any file. I'm following the below approach to pass the `cb: (err?: any) => void` parameter down to each of the replace file methods. But I'm not sure how to break the command execution using `cb(new CommandError(error))` in the `catch` block for these methods.

```
public commandAction(cmd: CommandInstance, args: CommandArgs, cb: (err?: any) => void): void {
  .
  .
  .
  this.replacePackageJsonContent(path.join(this.projectRootPath, 'package.json'), args, cmd, cb);
  this.replaceYoRcJsonContent(path.join(this.projectRootPath, '.yo-rc.json'), newId, args, cmd, cb);
  this.replacePackageSolutionJsonContent(path.join(this.projectRootPath, 'config', 'package-solution.json'), projectName, newId, args, cmd, cb);
  this.replaceDeployAzureStorageJsonContent(path.join(this.projectRootPath, 'config', 'deploy-azure-storage.json'), args, cmd, cb);
  this.replaceReadMeContent(path.join(this.projectRootPath, 'README.md'), projectName, args, cmd, cb);

  cb();
}

private replacePackageJsonContent = (filePath: string, args: CommandArgs, cmd: CommandInstance, cb: (err?: any) => void) => {
  try {
    .
    .
    .
  } catch (error) {
    cb(new CommandError(error));
  }
}

//other similar replace file methods
.
.
.
```

waldekmastykarz commented:
The only way would be to throw an exception with the error message and catch it in the `commandAction` method where you'd call `cb(new CommandError(msg))` which you now do in `replacePackageJsonContent`.

aakashbhardwaj619 commented:
Hi @waldekmastykarz I've made the required changes and marked this PR for review. Let me know if any other changes are needed.

aakashbhardwaj619 commented:
Hi @waldekmastykarz I've updated my fork and pushed a fix for the issue.

The tests were failing due to the upgraded `uuid` version to v8. The way `uuid/v4` method, to generate new solution id, was being stubbed in earlier version is not supported now, so I have added a wrapper method `generateNewId` to call `v4` and I'm stubbing this in the test file.

waldekmastykarz commented:
Thank you for the quick turnaround 👏

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -0,0 +1,247 @@
+      updatedContent.name = args.options.newName;
+    }
+    if (JSON.stringify(updatedContent, null, 2) !== JSON.stringify(existingContent, null, 2)) {
+      fs.writeFileSync(filePath, JSON.stringify(updatedContent, null, 2), 'utf-8');
```
Since we stringified the object previously, we should reuse it rather than stringifying it again

aakashbhardwaj619 replied:
Agreed. Updated now.

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -0,0 +1,247 @@
+      updatedContent.name) {
+      updatedContent.name = args.options.newName;
+    }
+    if (JSON.stringify(updatedContent, null, 2) !== JSON.stringify(existingContent, null, 2)) {
```
Isn't existingContent a string already (see line 123)?

aakashbhardwaj619 replied:
Yes @waldekmastykarz, sorry about that. Not sure how I missed that. Updated now.

waldekmastykarz commented:
Appreciate the quick turnaround ❤️

waldekmastykarz commented on src/o365/spfx/commands/project/project-rename.ts:
```diff
@@ -0,0 +1,251 @@
+    const chalk = vorpal.chalk;
+    log(vorpal.find(commands.PROJECT_RENAME).helpInformation());
+    log(
+      `   ${chalk.yellow('Important:')} Run this command in the folder where the project for which you
```
Seems like a remainder from the original command ;)

waldekmastykarz commented:
Merged manually. Nicely done! 👏
